### PR TITLE
flags: add new --disable flags to reduce noise and disable backgrounding

### PR DESCRIPTION
**Pull request by paulirish**
described mostly here: https://blog.bigbinary.com/2018/08/15/debugging-failing-tests-in-background-tab-in-puppeteer.html

also relevant: https://github.com/GoogleChrome/puppeteer/blame/master/lib/Launcher.js

test page for occlusion and timer drift: https://output.jsbin.com/goduxox/1/quiet

ref #169

### CODE REVIEW

File: docs/chrome-flags-for-tools.md [hunk 1]
```diff
@@ -94,6 +94,10 @@ These flags are being used in various tools. They also just need to be documente
 --disable-features=site-per-process # Disables OOPIF. https://www.chromium.org/Home/chromium-security/site-isolation
 --disable-hang-monitor
 
+--disable-backgrounding-occluded-windows
```
Comment (patrickhulce) [inline docs/chrome-flags-for-tools.md]: seems like it shouldn't be in both lists?

File: docs/chrome-flags-for-tools.md [hunk 2]
```diff
@@ -94,6 +94,10 @@ These flags are being used in various tools. They also just need to be documente
 --disable-features=site-per-process # Disables OOPIF. https://www.chromium.org/Home/chromium-security/site-isolation
```
Comment (patrickhulce) [inline docs/chrome-flags-for-tools.md]: add --enable-automation to this list?

**Comment by brendankenny** [comment]
> fixes #169

seems like the problem wasn't fixed by adding those flags in that issue? I'm confused :)

File: src/flags.ts [hunk 3]
```diff
@@ -10,6 +10,8 @@ export const DEFAULT_FLAGS: ReadonlyArray<string> = [
   '--disable-features=TranslateUI',
   // Disable all chrome extensions entirely
   '--disable-extensions',
+  // Disable some extensions that aren't affected by --disable-extensions
```
Comment (brendankenny) [inline src/flags.ts]: should probably remove `entirely` from L11 then :)

File: docs/chrome-flags-for-tools.md (hunk 2 above)
Comment (paulirish) [reply docs/chrome-flags-for-tools.md]: in fact we have the best docs for that particular flag. ;) https://github.com/GoogleChrome/chrome-launcher/blob/master/docs/chrome-flags-for-tools.md#--enable-automation

File: docs/chrome-flags-for-tools.md (hunk 1 above)
Comment (paulirish) [reply docs/chrome-flags-for-tools.md]: the md is just for other tool developers and hackers like vikram who are digging into this. ;)

**Comment by paulirish** [comment]
> Side effect of making GoogleChrome/lighthouse#9650 difficult to test?

yup. will have to relaunch without that flag in order to test. oh boy.

File: docs/chrome-flags-for-tools.md (hunk 1 above)
Comment (patrickhulce) [reply docs/chrome-flags-for-tools.md]: oh I see now :)

File: docs/chrome-flags-for-tools.md (hunk 2 above)
Comment (patrickhulce) [reply docs/chrome-flags-for-tools.md]: oh haha nice, I wish GH had a ctrl+f for the whole file easily available without having to manually expand the undiff'd parts of it

**Comment by VikramTiwari** [comment]
Just wanted to chime in and [confirm](https://github.com/GoogleChrome/chrome-launcher/pull/170#issuecomment-530137047) that even after adding these flags the pages in the backgrounded tab were throttled. 

Though, I am not sure if it was just the render, timer or actual node code running through puppeteer that was throttled. I can do more digging if it's unknown/helpful.

**Comment by paulirish** [comment]
@VikramTiwari thanks. that's useful. do you have a minimal test page we can experiment with?

also fwiw here are all the chromium flags: https://peter.sh/experiments/chromium-command-line-switches/

**Comment by VikramTiwari** [comment]
I was trying this on my [LinkedIn profile page](https://www.linkedin.com/in/vikramtiwari/) but you should be able to do it on any profile. Actions were basic copying and pasting.

Minimal script would:
- Load the profile page in a new page
- Use puppeteer's `page.bringToFront()` to bring some other page to front (easiest is `pages[0]`)
- perform copy operations on the older page

Best would be to add multiple copy operations so that you can see the throttle happening. I will get started on creating this into a gist that we can act as proof.

**Comment by VikramTiwari** [comment]
Seems like I am unable to reproduce the issue in my [POC](https://gist.github.com/VikramTiwari/77d0b81e605b0af161e0021e130d34d3) when I use the required flags. I can recreate it in my actual project though. So most probably it's on my side. Will update on the issue if I find something.

**Comment by VikramTiwari** [comment]
Okay! I can reproduce the issue now. Seems like it's just `mousemove`s which cause the slow down to happen. I have updated the POC with random mousemoves between copy events. If you comment [randomMouseMoves](https://gist.github.com/VikramTiwari/77d0b81e605b0af161e0021e130d34d3#file-index-js-L45) the execution runs as normal expected speed.

**Comment by paulirish** [comment]
https://output.jsbin.com/goduxox/1/quiet shows that with these flags....

- timers are no longer throttled when occluded or in a background tab
- the page visibility api/events are unaffected (it still works)
- renderer painting still works when the window is occluded
- however renderer painting doesn't work when the tab is backgrounded.

So this won't fix cli NO_FCP issues entirely but will help.

**Comment by paulirish** [comment]
I'm ready to land this (as soon at CI is done)

**Pull request merged**